File: apps/backend-python/app/modules/payment/infrastructure/persistence/mappers/payment_mapper.py
"""
Payment Mapper

Mapper для конвертации между Payment Entity и PaymentModel.
"""
from app.modules.payment.domain import (
    Payment,
    PaymentStatus,
    PaymentMethod,
    Money,
    RefundReason,
)
from app.modules.payment.infrastructure.persistence.models import PaymentModel


class PaymentMapper:
    """
    Mapper для конвертации Payment Entity <-> PaymentModel.
    """

    @staticmethod
    def to_domain(model: PaymentModel) -> Payment:
        """
        Конвертирует ORM модель в Domain Entity.
        """
        # Создаем Value Objects
        status = PaymentStatus(value=model.status)
        payment_method = PaymentMethod(value=model.payment_method)
        amount = Money(amount=model.amount, currency=model.currency)

        # Refund amount
        refund_amount = None
        if model.refund_amount:
            refund_amount = Money(amount=model.refund_amount, currency=model.currency)

        # Refund reason
        refund_reason = None
        if model.refund_reason:
            refund_reason = RefundReason(
                value=model.refund_reason,
                comment=model.refund_reason_comment or "",
            )

        # Создаем Entity через __new__ (восстановление из БД)
        payment = Payment.__new__(Payment)
        payment._id = model.id
        payment._user_id = model.user_id
        payment._consultation_id = model.consultation_id
        payment._subscription_id = model.subscription_id
        payment._amount = amount
        payment._status = status
        payment._payment_method = payment_method
        payment._external_payment_id = model.external_payment_id
        payment._failure_reason = model.failure_reason
        payment._refund_reason = refund_reason
        payment._refund_amount = refund_amount
        payment._processed_at = model.processed_at
        payment._refunded_at = model.refunded_at
        payment._metadata = model.metadata or {}
        payment._created_at = model.created_at
        payment._updated_at = model.updated_at
        payment._domain_events = []

        return payment

    @staticmethod
    def to_model(entity: Payment) -> PaymentModel:
        """
        Конвертирует Domain Entity в ORM модель.
        """
        return PaymentModel(
            id=entity.id,
            user_id=entity.user_id,
            consultation_id=entity.consultation_id,
            subscription_id=entity.subscription_id,
            amount=entity.amount.amount,
            currency=entity.amount.currency,
            status=entity.status.value,
            payment_method=entity.payment_method.value,
            external_payment_id=entity.external_payment_id,
            failure_reason=entity.failure_reason,
            refund_amount=entity.refund_amount.amount if entity.refund_amount else None,
            refund_reason=entity.refund_reason.value if entity.refund_reason else None,
            refund_reason_comment=entity.refund_reason.comment if entity.refund_reason else None,
            metadata=entity.metadata,
            processed_at=entity.processed_at,
            refunded_at=entity.refunded_at,
            created_at=entity.created_at,
            updated_at=entity.updated_at,
        )

    @staticmethod
    def update_model(entity: Payment, model: PaymentModel) -> None:
        """
        Обновляет существующую ORM модель данными из Entity.
        """
        model.status = entity.status.value
        model.external_payment_id = entity.external_payment_id
        model.failure_reason = entity.failure_reason
        model.refund_amount = entity.refund_amount.amount if entity.refund_amount else None
        model.refund_reason = entity.refund_reason.value if entity.refund_reason else None
        model.refund_reason_comment = entity.refund_reason.comment if entity.refund_reason else None
        model.processed_at = entity.processed_at
        model.refunded_at = entity.refunded_at
        model.metadata = entity.metadata
        model.updated_at = entity.updated_at

File: apps/backend/src/modules/payment/infrastructure/persistence/payment.mapper.ts
import { Payment } from '../../domain/entities/payment.entity';
import { Money } from '../../domain/value-objects/money.vo';
import { PaymentStatus } from '../../domain/value-objects/payment-status.vo';
import { PaymentMethod } from '../../domain/value-objects/payment-method.vo';
import { PaymentOrmEntity } from './payment.orm-entity';

/**
 * Payment Mapper
 *
 * Maps between Payment domain entity and PaymentOrmEntity (persistence).
 * Follows the Repository pattern for clean architecture.
 */
export class PaymentMapper {
  /**
   * Map ORM entity to Domain entity
   */
  public static toDomain(ormEntity: PaymentOrmEntity): Payment {
    // Create Money value object
    const moneyResult = Money.create({
      amount: Number(ormEntity.amount),
      currency: ormEntity.currency,
    });

    if (moneyResult.isFailure) {
      throw new Error(
        `Failed to map ORM entity to domain: ${moneyResult.error}`,
      );
    }

    const money = moneyResult.getValue();

    // Create refunded amount if present
    let refundedMoney: Money | undefined;
    if (ormEntity.refundedAmount) {
      const refundedMoneyResult = Money.create({
        amount: Number(ormEntity.refundedAmount),
        currency: ormEntity.currency,
      });

      if (refundedMoneyResult.isSuccess) {
        refundedMoney = refundedMoneyResult.getValue();
      }
    }

    // Create payment with minimal required data
    const paymentResult = Payment.create(
      ormEntity.id,
      ormEntity.userId,
      money,
      ormEntity.description,
      ormEntity.consultationId,
      ormEntity.subscriptionId,
      ormEntity.metadata,
    );

    if (paymentResult.isFailure) {
      throw new Error(
        `Failed to map ORM entity to domain: ${paymentResult.error}`,
      );
    }

    const payment = paymentResult.getValue();

    // Set additional properties that aren't set in create()
    (payment as any).props.status = ormEntity.status as PaymentStatus;
    (payment as any).props.method = ormEntity.method as PaymentMethod;
    (payment as any).props.yooKassaPaymentId = ormEntity.yooKassaPaymentId;
    (payment as any).props.yooKassaPaymentUrl = ormEntity.yooKassaPaymentUrl;
    (payment as any).props.refundedAmount = refundedMoney;
    (payment as any).props.failureReason = ormEntity.failureReason;
    (payment as any).props.createdAt = ormEntity.createdAt;
    (payment as any).props.updatedAt = ormEntity.updatedAt;
    (payment as any).props.completedAt = ormEntity.completedAt;
    (payment as any).props.canceledAt = ormEntity.canceledAt;
    (payment as any).props.refundedAt = ormEntity.refundedAt;

    return payment;
  }

  /**
   * Map Domain entity to ORM entity
   */
  public static toOrm(domainEntity: Payment): PaymentOrmEntity {
    const ormEntity = new PaymentOrmEntity();

    ormEntity.id = domainEntity.id;
    ormEntity.userId = domainEntity.userId;
    ormEntity.consultationId = domainEntity.consultationId;
    ormEntity.subscriptionId = domainEntity.subscriptionId;
    ormEntity.amount = domainEntity.amount.amount;
    ormEntity.currency = domainEntity.amount.currency;
    ormEntity.status = domainEntity.status;
    ormEntity.method = domainEntity.method;
    ormEntity.description = domainEntity.description;
    ormEntity.yooKassaPaymentId = domainEntity.yooKassaPaymentId;
    ormEntity.yooKassaPaymentUrl = domainEntity.yooKassaPaymentUrl;
    ormEntity.refundedAmount = domainEntity.refundedAmount?.amount;
    ormEntity.failureReason = domainEntity.failureReason;
    ormEntity.metadata = domainEntity.metadata;
    ormEntity.createdAt = domainEntity.createdAt;
    ormEntity.updatedAt = domainEntity.updatedAt;
    ormEntity.completedAt = domainEntity.completedAt;
    ormEntity.canceledAt = domainEntity.canceledAt;
    ormEntity.refundedAt = domainEntity.refundedAt;

    return ormEntity;
  }

  /**
   * Map multiple ORM entities to Domain entities
   */
  public static toDomainMany(ormEntities: PaymentOrmEntity[]): Payment[] {
    return ormEntities.map((ormEntity) => this.toDomain(ormEntity));
  }

  /**
   * Map multiple Domain entities to ORM entities
   */
  public static toOrmMany(domainEntities: Payment[]): PaymentOrmEntity[] {
    return domainEntities.map((domainEntity) => this.toOrm(domainEntity));
  }
}
